Dashboard Improvements: Add Opening Times, Improve UX/UI

sagzy opened the pull request:
This PR:
- Improves the UX/UI first version of the dashboard (with e.g. a single address field that is autofilled by Mapbox Geocoder when adding / editing a location)
- Add opening times
- Fixes a couple of other things (authorization, UI of the flash message, UI of the login / signup, small details in the UX/UI here and there)

shafy commented on app/javascript/controllers/opening_time_controller.js:
```diff
@@ -0,0 +1,82 @@
+  ]
+
+  initialize() {
+    this.toggleClosed = this.toggleClosed.bind(this)
+    this.toggleOpen24h = this.toggleOpen24h.bind(this)
```
Why do you need to bind it?

sagzy replied:
To be able to refer to the Stimulus controller in the `toggleClosed` / `toggleOpen24h` methods. Without it, we can't call `this.timesBlockTarget` etc.

shafy commented on app/javascript/controllers/opening_time_controller.js:
```diff
@@ -0,0 +1,82 @@
+    if (!this.closedTarget) return
+    if (!this.open24hTarget) return
+
+    this.closedTarget.addEventListener('change', this.toggleClosed)
+    this.open24hTarget.addEventListener('change', this.toggleOpen24h)
```
Target Callbacks coming soon :D https://github.com/hotwired/stimulus/issues/445

sagzy replied:
Nice idea!

shafy commented on app/views/development/design/index.html.erb:
```diff
@@ -6,17 +6,34 @@
 <div class="min-w-0 flex-1 grid grid-cols-5 2xl:grid-cols-10 gap-x-4 gap-y-3 2xl:gap-x-2">
   <div class="space-y-1.5">
-    <div class="h-10 w-full rounded ring-1 ring-inset ring-black ring-opacity-0" style="background-color:#4338ca"></div>
+    <div class="h-10 w-full rounded ring-1 ring-inset ring-black ring-opacity-0" style="background-color:#704DCC"></div>
```
Oh nice, didn't know `background-color:#704DCC` worked.

shafy commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -4,38 +4,38 @@
     <%= render "devise/shared/error_messages", resource: resource %>
     <div class="field mt-4">
-      <%= f.label :full_name, class: "block text-gray-700" %>
+      <%= f.label :full_name, class: 'block text-gray-700' %>
```
Why change to single quote in this file?

sagzy replied:
good one! fixed

shafy commented on app/views/shared/_location.html.erb:
```diff
@@ -1,16 +1,16 @@
+      <% @location.opening_times.each do |ot| %>
+        🕛
+        <%= ot.to_s %>
+        <br />
```
I think th HTML syntax is `<br>`, `<br />` works in HTML too but comes from XHTML. Not important, just fyi!

sagzy replied:
Ah, TIL :D Let me change these babies

sagzy replied:
✅

shafy commented on db/seeds.rb:
```diff
@@ -1,179 +1,35 @@
+
+# Map 3: Zürich
+
+user = User.create!(name: "Tyrion Lannister", email: "tyrion@mapzy.io", password: "smartass007", password_confirmation: "smartass007")
```
Uuuhh, welcome Tyrion!

sagzy replied:
🗡️

shafy commented on spec/workers/email_worker_spec.rb:
```diff
@@ -16,13 +16,13 @@
 
-    shared_examples "a correct address email" do |mail|
+    shared_examples "a correct address email" do |_mail|
       it "sends out email to correct address" do
         expect(@mail.to.first).to eq(user.email)
       end
     end
 
-    shared_examples "a correct subject email" do |mail, email_subject|
+    shared_examples "a correct subject email" do |_mail, email_subject|
       it "sends out email with correct subject" do
         expect(@mail.subject).to eq(email_subject)
       end
```
I know I wrote this, but I think it's better to pass the email as an argument instead of using the class variable. Could you update that?

sagzy replied:
Happy to, chief! But I'm not sure I got what you meant, could you clarify?

shafy reviewed (approved):
Looks great, just addd a few minor comments.

Also, in `database.yml` the `host:` values for `development` and `test` need to be:
`<%= ENV.fetch("DATABASE_HOST") { "localhost" } %>` instead of `<%= ENV.fetch("DATABASE_HOST") { localhost } %>`, otherwise it expects `localhost` to be a variable (which is of course not defined here).

I'll commit a fix right now.

I'll also play around a bit and see if I have some more comments. Stay tuned!

shafy commented:
First of all, FUCKING LOVE THE UX. Having one address line is a fucking game changer and feels so much better.

Here are some comments that I think we should change before merging:

- Do we need the full name when signing up? The fewer info to enter, the higher the conversion rate (and the fewer PII we have to worry about). I suggest to remove it.
- I suggest not to automatically hide the flash message for trial accounts. Give users time to read and dimiss themselves if they want to.
- Let's change the trial message to: "Thanks for trying out Mapzy. Your trial ends on Nov 13, 2021. Make sure to subscribe to one of our plans before that." (removing "and your map will be set to inactive", I think it's too brutal in-your-face right in the beginning").
- When first loading the map, it says `mapboxgl` is undefined and doesn't display the map. I think this is related to the fact that `map_controller.js` (or `location_controller.js`) is executed before the Mapbox library is loaded. I suggest to include the Mapbox library with [rails-importmap](https://github.com/rails/importmap-rails/) so that we can import it directly in the Stimulus controllers.
- Dragging the marker doesn't update the entered address. Which is fine, keep it simple. Maybe later on we could add a message if the dragged distance is far, saying "Heads up, your marker location might not be at the address you entered" or something.

sagzy commented on app/controllers/dashboard_controller.rb:
```diff
@@ -13,16 +12,23 @@ class DashboardController < ApplicationController
   def show_trial_reminder
     return unless current_user.account.trial?
-    flash[:notice] = trial_reminder_text
+    return if cookies[:trial_reminder_shown]
+
+    cookies[:trial_reminder_shown] = {
+      value: "true",
+      expires: 1.day
+    }
+
+    flash.now[:notice] = trial_reminder_text
```
I'd find that very annoying, at least in the current way we display the reminder (obstrusive flash message). Plus it kind of conflicts with the other flash messages we use (after a location creation, etc.). I think if we display the reminder once per day + all the reminder emails it's already informative enough. Don't want to be too spammy.

What do you think?

sagzy commented:
Thanks dude! Addressed most of your feedback, left some comments directly below.


>     * Do we need the full name when signing up? The fewer info to enter, the higher the conversion rate (and the fewer PII we have to worry about). I suggest to remove it.

Makes sense, removed the name on the signup, now there is only email & password. Also removed the "current password" in the account settings - it asks for it even if you just want to add your name, which I find overkill. Better UX without it. Let me know if you think it's too unsafe.

>     * I suggest not to automatically hide the flash message for trial accounts. Give users time to read and dimiss themselves if they want to.

I've increased it to 30 sec now (was 15 sec before). I think it's better not to be too annoying with the trial reminder, we also show it everyday + the user get a couple of emails during the trial period. I think if the user want to subscribe, they'll find where it is (go to settings and it's the first thing shown). Wouldn't be too spammy about it. You disagree?

>     * Let's change the trial message to: "Thanks for trying out Mapzy. Your trial ends on Nov 13, 2021. Make sure to subscribe to one of our plans before that." (removing "and your map will be set to inactive", I think it's too brutal in-your-face right in the beginning").

Good idea, changed

>     * When first loading the map, it says `mapboxgl` is undefined and doesn't display the map. I think this is related to the fact that `map_controller.js` (or `location_controller.js`) is executed before the Mapbox library is loaded. I suggest to include the Mapbox library with [rails-importmap](https://github.com/rails/importmap-rails/) so that we can import it directly in the Stimulus controllers.

Ah really? I don't get this. Using importmap is a good idea anyways, let me look into how it works

>     * Dragging the marker doesn't update the entered address. Which is fine, keep it simple. Maybe later on we could add a message if the dragged distance is far, saying "Heads up, your marker location might not be at the address you entered" or something.

Yeah good idea too, let's keep it for later. I think most users won't need to drag the marker anyway and for those who do, the max zoom in on the marker should already avoid this kind of scenario. But would be a good addition to the UX, the more informative the better.

shafy commented:
@sagzy perfect, thanks! After adding importmaps this ready to merge from my point of view.

sagzy commented:
Yes and fixing the specs and linting :) 

> Le 31 oct. 2021 à 12:35, Can Olcer ***@***.***> a écrit :
> 
> @sagzy perfect, thanks! After adding importmaps this ready to merge from my point of view.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

shafy commented:
@sagzy I have removed again the no password requirement because [that also allows users to change their password without needing in to put a matching password](https://github.com/heartcombo/devise/wiki/How-To:-Allow-users-to-edit-their-account-without-providing-a-password). I think it's better to keep the current password requirement for now (there's a way to make it work also without it, but I feel like it's not worth getting into it now).

The failing capybara tests are because of the selenium webdriver I think. We need to configure GitHub Actions so that it runs the headless Chrome browser. Can you do that?
